File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/MetalBorrowedTextureDescriptor.java
package org.maplibre.nativeffi;

import java.util.Objects;

/** Mutable descriptor for Metal caller-owned texture render targets. */
public final class MetalBorrowedTextureDescriptor {
  private int width = 256;
  private int height = 256;
  private double scaleFactor = 1.0;
  private NativePointer texture = NativePointer.NULL;

  public int width() {
    return width;
  }

  public int height() {
    return height;
  }

  public MetalBorrowedTextureDescriptor setSize(int width, int height) {
    if (width <= 0 || height <= 0) {
      throw new IllegalArgumentException("width and height must be positive");
    }
    this.width = width;
    this.height = height;
    return this;
  }

  public double scaleFactor() {
    return scaleFactor;
  }

  public MetalBorrowedTextureDescriptor setScaleFactor(double scaleFactor) {
    if (!Double.isFinite(scaleFactor) || scaleFactor <= 0.0) {
      throw new IllegalArgumentException("scaleFactor must be finite and positive");
    }
    this.scaleFactor = scaleFactor;
    return this;
  }

  public NativePointer texture() {
    return texture;
  }

  public MetalBorrowedTextureDescriptor setTexture(NativePointer texture) {
    this.texture = Objects.requireNonNull(texture, "texture");
    return this;
  }
}

File: bindings/kotlin/src/commonMain/kotlin/org/maplibre/nativeffi/render/MetalBorrowedTextureDescriptor.kt
package org.maplibre.nativeffi.render

import org.maplibre.nativeffi.internal.status.Status

/** Mutable descriptor for Metal caller-owned texture render targets. */
public class MetalBorrowedTextureDescriptor(
  extent: RenderTargetExtent,
  physicalWidth: Int,
  physicalHeight: Int,
  texture: NativePointer,
) {
  public var extent: RenderTargetExtent = extent

  /** Physical texture size in device pixels, stated by the texture's owner rather than derived. */
  public var physicalWidth: Int = physicalWidth
    set(value) {
      Status.requireArgument(value >= 0) { "physicalWidth must be non-negative" }
      field = value
    }

  /** Physical texture size in device pixels, stated by the texture's owner rather than derived. */
  public var physicalHeight: Int = physicalHeight
    set(value) {
      Status.requireArgument(value >= 0) { "physicalHeight must be non-negative" }
      field = value
    }

  public var texture: NativePointer = texture

  init {
    Status.requireArgument(physicalWidth >= 0) { "physicalWidth must be non-negative" }
    Status.requireArgument(physicalHeight >= 0) { "physicalHeight must be non-negative" }
  }
}
